gcc/rust/expand/rust-macro-substitute-ctx.cc

```diff
@@ -31,7 +31,13 @@ SubstituteCtx::substitute_metavar (
   if (it == fragments.end ())
     {
       // fail to substitute
-      return false;
+
+      // HACK: substitute ($ crate) => (crate)
+      if (metavar->get_id () != CRATE)
+	return false;
+
+      expanded.push_back (metavar->clone_token ());
+      return true;
```

Comment:
I'm not sure this is enough to get the proper behavior of `$crate`, and we will run into issues when importing the `core` crate if I'm not mistaken. For example, when expanding [`panic!()`](https://doc.rust-lang.org/src/std/macros.rs.html#14-20), we need to make sure we're calling into `core`'s `panic_2015!` or `panic_2018!` macro and not a similarly named macro defined by the user in the crate where `panic!` is being *called*. Unless I am misunderstanding?

Reply:
Yeah, ```$crate``` would need to be handled better for macro exporting purposes. As long as a macro isn't exported, ```$crate``` => ```crate``` should still work. I figured a partial fix to get ```libcore``` parsing would still be useful, and of course this shouldn't resolve #1078.